---
title: "Single-cell RNAseq analysis of the human melanoma cells | IBDM | Marie-Claire Delfini"
subtitle: "Seurat | QC and Clustering"
author:
- Thomas Vannier | CENTURI Multi-Engineering Platform
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_document: 
    theme: cerulean
    toc: true
    toc_float: true
    self_contained: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error=TRUE)
```

```{r loadLibraries, echo=FALSE, warning=FALSE, message=FALSE}
library(rmarkdown)
library(tidyverse)
library(stringr)
library(tinytex)
library(knitr)
library(readr)
library(dplyr)
library(Seurat)
library(patchwork)
library(sctransform)
library(glmGamPoi)
library(ggplot2)
library(data.table)
library(devtools)
library(sessioninfo)
library(dittoSeq)
```

[//]: # "Load the various chunks in the various files"
```{r readChunks, echo=FALSE, warning=FALSE, message=FALSE}
read_chunk( path=file.path( SCRIPTDIR, "assignment.R"))
read_chunk( path=file.path( SCRIPTDIR, "seurat.R"))
```
This report present the analyse of two dataset generated with 10x Genomics Single Cell Protocol:

- Wouters et al., 2019:  scRNA-seq on nine patient-derived 2D melanoma cultures (MM lines), as well as the A375 melanoma cell line. 

  3 Melanocyte-like (MM001, MM011, MM031 -> SOX10-high melanocyte-like state)

  3 Intermediate (MM074, MM087, MM057 -> NFAT2C)

  3 mesenchymal-like (MM029, MM047, MM099 -> SOX9-high mesenchymal-like state)

  A375 melanoma cell line.

- Pietrobono et al., 2019: Four cell types were used for single-cell RNA sequencing. Normal human neonatal epidermal melanocytes. Melanoma patient-derived xenograft lines M12, M15, and M27.

# QC and selecting cells for further analysis of the Projects `r snakemake@params[["biop"]]`

## Visualize QC metrics as a violin plot and FeatureScatter

We visualize QC metrics, and use these to filter cells.

- We filter cells that have unique feature counts over `r snakemake@params[["maxFeature_RNA"]]` or less than `r snakemake@params[["minFeature_RNA"]]`

- We filter cells that have >`r snakemake@params[["percent_mt"]]`% mitochondrial counts.

```{r qcmetrics, echo=FALSE,fig.height=10,fig.width=10,out.height=600,out.width=600, results = "asis", message=FALSE}
<<qcmetrics>>
```

## Normalizing the data

After removing unwanted cells from the dataset we normalize the feature expression measurements for each cell (and correct the batch effect using the Seurat integration procedure if necessary) with `r snakemake@params[["normwf"]]`
We apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction techniques like PCA. If SCTransform is performed, it's not necessary to scale the data after it.

## Identification of highly variable features

Subset of features that exhibit high cell-to-cell variation in the dataset

```{r hvf, echo=FALSE,fig.height=10,fig.width=8,out.height=600,out.width=450, results = "asis", message=FALSE, warning=FALSE}
<<hvf>>
```

## Perform linear dimensional reduction

Visualizing both cells and features that define the PCA

```{r ldr, echo=FALSE,fig.height=10,fig.width=10,out.height=600,out.width=600, results = "asis", message=FALSE}
<<ldr>>
```

## Determine the ‘dimensionality’ of the dataset

How many components have to be include?

### Elbow Plot

A ranking of principle components based on the percentage of variance explained by each one.

```{r elbowplot, echo=FALSE, message=FALSE}
<<elbowplot>>
```

## Cluster the cells

We use a graph-based clustering approach :
1. Construct a KNN graph based on the euclidian distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity).
2. Modularity optimization techniques such as the Louvain algorithm or SLM, to iteratively group cells together, with the goal of optimizing the standard modularity function. 

```{r cluster, echo=FALSE, fig.height=10, fig.width=10, out.height=600, out.width=600, results = "asis", message=FALSE, eval = TRUE}
<<cluster>>
```

## Run non-linear dimensional reduction

UMAP to visualize and explore these datasets.

```{r umap, echo=FALSE, fig.height=10, fig.width=10, out.height=600, out.width=600, results = "asis", message=FALSE, warning=FALSE, eval = TRUE}
<<umap>>
```

## Retrieve data in matrix

Count, expression and normalized matrice can be found in the snakemake report

```{r datasave, echo=FALSE, message=FALSE, eval = TRUE}
<<datasave>>
```

## Session Info

```{r info, echo=FALSE, message=FALSE, eval = TRUE}
<<info>>
```